ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision/BulkSync.hs

```diff
@@ -0,0 +1,591 @@
+--     add all of @inflight(currentPeer)@ to @inflightIgnored@. REVIEW: Nick's
+--     description says to add all of @inflight@ (for all the peers) to
+--     @inflightIgnored@, probably because it assumes/enforces that @inflight ==
+--     inflight(currentPeer)@, that is only the current peer is allowed to have
+--     in-flight blocks.
+--
+--   - Stop considering the peer “current” and make them the worst according to
+--     the @peersOrder@.
+--
+-- - Let @theCandidate :: AnchoredFragment (Header blk)@ be the best candidate
+--   header chain among the ChainSync clients (eg. best raw tiebreaker among the
+--   longest).
+--
+-- - Let @grossRequest@ be the oldest blocks on @theCandidate@ that have not
+--   already been downloaded, are not in @inflightNonIgnored@, and total less
+--   than 20 mebibytes.
+--
+-- - If @grossRequest@ is empty, then terminate this iteration. Otherwise, pick
+--   the best peer (according to @peersOrder@) offering all of the blocks in
+--   @grossRequest@. We will call it @thePeer@. Because @currentPeer@, if it
+--   exists, is the best according to @peersOrder@, then it will be our
+--   preferred peer, as long as it can provide the @grossRequest@s.
+--
+-- - If the byte size of @inflight(thePeer)@ is below the low-water mark, then
+--   terminate this iteration. Otherwise, decide and send the actual next batch
+--   request, as influenced by exactly which blocks are actually already
+--   currently in-flight with the chosen peer.
+--
+-- - Update @currentPeer@ and @currentStart@, if needed. Namely:
+--
+--   - If @thePeer /= currentPeer@, then make @thePeer@ the current peer and the
+--     best according to @peersOrder@, and reset @currentStart@ to now.
+--
+--   - If @thePeer == currentPeer@, but @inflight(thePeer)@ is empty, then reset
+--     @currentStart@ to now.
+--
+-- Terminate this iteration.
+--
+-- Interactions with ChainSync Jumping (CSJ)
+-- -----------------------------------------
+--
+-- This decision logic is not so obviously coupled with CSJ, but it is in some
+-- subtle ways:
+--
+-- - Because we always require our peers to be able to serve a gross request of
+--   oldest blocks, peers with longer chains have a better chance to pass this
+--   criteria and to be selected as current peer. The CSJ dynamo, being always
+--   ahead of jumpers, has therefore more chances to be selected as the current
+--   peer. It is still possible for a jumper or a disengaged peer to be
+--   selected.
+--
+-- - If the current peer is the CSJ dynamo, but it is a dishonest peer serving
+--   headers fast but retaining blocks, it might be able to drastically leash
+--   us, because its ChainSync client will be stuck behind the forecast horizon
+--   (and therefore not subject to ChainSync punishments such as the Limit on
+--   Patience). This is why we need to consider starvation of ChainSel and
+--   demote peers that let us starve.
```

Comment:
I don't understand this bit yet. If the dynamo is blocked on the forecast horizon, BlockFetch will still download blocks from some other peer because ChainSel will starve just the same.

What need is there to rotate the dynamos?

Reply:
I think @nfrisby is better equipped to answer this.